src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java

```diff
@@ -408,12 +418,32 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
 		return ret;
 	}
 
+	private static int getMrjVersionFromPath(Path path) {
+		if (File.separator.equals("/")) {
+			return getMrjVersionFromPath(path.toString());
+		} else if (File.separator.equals("\\")) {
+			return getMrjVersionFromPath(path.toString().replace('\\', '/'));
+		} else {
+			throw new RuntimeException("Unknown file separator detected.");
+		}
+	}
+
+	private static int getMrjVersionFromPath(String str) {
+		Pattern pattern = Pattern.compile("(?<=" + ClassInstance.MRJ_PREFIX + "/)[0-9]*");
+		Matcher matcher = pattern.matcher(str);
+		if (matcher.find()) {
+			return Integer.parseInt(matcher.group());
+		}
+		return ClassInstance.MRJ_DEFAULT;
```

Comment:
I don't really like how this just searches the path string, it should be possible to determine this accurately due to the class name <-> path relationship in Java.

I'd do the following:
- instead of passing `int mrjVersion` and `byte[] data` to `analyze()`, pass the `file` and whether it's a .jar or .class
- do the `Files.readAllBytes` in `analyze()`
- set the mrjVersion in `ClassInstance.init()` instead of the constructor
- determine said mrjVersion just before the `ClassInstance.init()` call in `visit()` using the class name and `file`

The last step works as follows for plain classes:
1. ensure the file path is absolute and normalized by making it so all the way back in `net.fabricmc.tinyremapper.TinyRemapper.read(Path, boolean, InputTag[], Path, boolean, List<FileSystem>)`
2. strip the class name parts from `file`, so `/path/to/some/pkg/TheClass.class` for `some.okg.TheClass` becomes `/path/to`
3. the parent dir of the path is now either null, the mrj version or something else
4. the parent dir of the parent dir is now versions for mrj, and the parent of that META-INF

For jars replace 2.-4. with checking the 1st component of the path for META-INF and the 2nd for the versions, the 3rd is the actual version

Reply:
fixed